### PR TITLE
feat(firebase_storage): add support for the Firebase Storage emulator via `useEmulator`

### PR DESCRIPTION
**Pull request by russellwheatley**
## Description

Initial work for Storage emulator on the web platform.

## Related Issues

none.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`).
This will ensure a smooth and quick review process. Updating the `pubspec.yaml` and changelogs is not required.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`melos run analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/FirebaseExtended/flutterfire/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

**Comment by Salakar** [comment]
Blocked pending iOS SDK 8 landing.

**Comment by Ehesp** [comment]
@googlebot I consent.

File: packages/firebase_storage/firebase_storage/test/mock.dart [hunk 1]
```diff
@@ -120,7 +127,7 @@ class TestFirebaseStoragePlatform extends FirebaseStoragePlatform {
 // ReferencePlatform Mock
 class TestReferencePlatform extends ReferencePlatform {
   TestReferencePlatform() : super(TestFirebaseStoragePlatform(), testFullPath);
-  // @override
+// @override
```
Comment (Salakar) [inline packages/firebase_storage/firebase_storage/test/mock.dart:130]: Remove?

File: packages/firebase_storage/firebase_storage_platform_interface/lib/src/platform_interface/platform_interface_firebase_storage.dart [hunk 2]
```diff
@@ -92,6 +92,17 @@ abstract class FirebaseStoragePlatform extends PlatformInterface {
     throw UnimplementedError('ref() is not implemented');
   }
 
+  /// Changes this instance to point to a Storage emulator running locally.
+  ///
+  /// Set the [host] and [port] of the local emulator, such as "http://localhost"
```
Comment (Salakar) [inline packages/firebase_storage/firebase_storage_platform_interface/lib/src/platform_interface/platform_interface_firebase_storage.dart:97]:
```suggestion
  /// Set the [host] and [port] of the local emulator, such as "localhost"
```

File: packages/firebase_storage/firebase_storage_web/lib/src/firebase_storage_web.dart [hunk 3]
```diff
@@ -118,4 +118,12 @@ class FirebaseStorageWeb extends FirebaseStoragePlatform {
   void setMaxDownloadRetryTime(int time) {
     _maxDownloadRetryTime = time;
   }
+
+  @override
+  Future<void> useEmulator(String host, int port) async {
+    // The generic platform interface is with host and port split to
+    // centralize logic between android/ios native, but web takes the
+    // origin as a single string
+    guard(() => webStorage!.useEmulator(host, port));
```
Comment (Salakar) [inline packages/firebase_storage/firebase_storage_web/lib/src/firebase_storage_web.dart:127]:
```suggestion
    guard(() => webStorage!.useEmulator(host, port));
```

It doesn't?

**Review by Salakar — CHANGES_REQUESTED** [review]
Minor points above

**Comment by Ehesp** [comment]
Update:

- Emulator reports wrong error when a ref is not found AND not readable.
- `read` rule on emulator isn't working.
- Update metadata is throwing a dart error on web.

File: .github/workflows/scripts/firebase.json [hunk 4]
```diff
@@ -12,9 +13,15 @@
     "auth": {
       "port": "9099"
     },
+    "storage": {
+      "port": "9080"
```
Comment (mikehardy) [inline .github/workflows/scripts/firebase.json:17]: Default port for storage emulator is 9199 https://github.com/firebase/firebase-js-sdk/blob/50860d794bfe711143b5d287e1042eeb5187e47b/packages/rules-unit-testing/src/api/index.ts#L48
```suggestion
      "port": "9199"
```

File: packages/firebase_storage/firebase_storage/example/test_driver/test_utils.dart [hunk 5]
```diff
@@ -14,6 +14,15 @@ Random _random = Random();
...
     length, (_) => _chars.codeUnitAt(_random.nextInt(_chars.length))));
 
+String get testEmulatorHost {
+  if (defaultTargetPlatform == TargetPlatform.android && !kIsWeb) {
+    return '10.0.2.2';
+  }
+  return 'localhost';
+}
+
+const int testEmulatorPort = 9080;
```
Comment (mikehardy) [inline packages/firebase_storage/firebase_storage/example/test_driver/test_utils.dart:24]: use standard / default storage emulator port

```suggestion
const int testEmulatorPort = 9199;
```

File: packages/firebase_storage/firebase_storage/lib/src/firebase_storage.dart [hunk 6]
```diff
@@ -149,6 +149,33 @@ class FirebaseStorage extends FirebasePluginPlatform {
     return _delegate.setMaxDownloadRetryTime(time.inMilliseconds);
   }
 
+  /// Changes this instance to point to a Storage emulator running locally.
+  ///
+  /// Set the [host] (ex: localhost) and [port] (ex: 5001) of the local emulator.
```
Comment (mikehardy) [inline packages/firebase_storage/firebase_storage/lib/src/firebase_storage.dart:154]: standard storage emulator port 

```suggestion
  /// Set the [host] (ex: localhost) and [port] (ex: 9199) of the local emulator.
```

File: packages/firebase_storage/firebase_storage_platform_interface/lib/src/full_metadata.dart [hunk 7]
```diff
@@ -44,6 +44,7 @@ class FullMetadata {
 
   /// Custom metadata set on this storage object.
   Map<String, String>? get customMetadata {
+    print('customMetadata ${_metadata}');
```
Comment (mikehardy) [inline packages/firebase_storage/firebase_storage_platform_interface/lib/src/full_metadata.dart:47]: debugging left over?

File: packages/firebase_storage/firebase_storage_web/lib/src/interop/storage.dart [hunk 8]
```diff
@@ -192,9 +198,11 @@ class StorageReference
 
   /// Updates metadata from this reference at actual location with
   /// the new [metadata].
-  Future<FullMetadata> updateMetadata(SettableMetadata metadata) =>
-      handleThenable(jsObject.updateMetadata(metadata.jsObject))
-          .then(FullMetadata.getInstance);
+  Future<FullMetadata> updateMetadata(SettableMetadata metadata) {
+    print('>>>> $metadata');
```
Comment (mikehardy) [inline packages/firebase_storage/firebase_storage_web/lib/src/interop/storage.dart:202]: debugging?

**Review by mikehardy — CHANGES_REQUESTED** [review]
mostly comments about using 9199 so the port is standard

**Comment by Ehesp** [comment]
Fixed the metadata issue, there was an implicit cast issue on web. Update the tests to work around emulator issues and commented out those which cannot be fixed until a fix lands on the emulator.

File: docs/storage/usage.mdx [hunk 9]
```diff
@@ -477,3 +477,14 @@ Future<void> handleTaskExample4(String filePath) async {
   }
 }
 ```
+
+## Emulator Usage
+
+If you are using the local [Storage emulators](https://firebase.google.com/docs/rules/emulator-setup), then it is possible to
```
Comment (Salakar) [inline docs/storage/usage.mdx:483]:
```suggestion
If you are using the local [Storage emulator](https://firebase.google.com/docs/rules/emulator-setup), then it is possible to
```

**Comment by Salakar** [comment]
@googlebot I consent.

**Pull request merged**